### PR TITLE
Fix referrer policy tests for a-tag

### PR DESCRIPTION
**Pull request by TheKK**
This RP tries to fix referrer policy test for <a> which was introduced in #12493 (sorry for my silly mistake). But the fact that Servo lakes of some functionalities make these tests a little tricky to do.

The desired solution for tests for <a> is to:
1. create a document which is running test harness
2. append an `<iframe>` to its parent document and give it a name
3. append and `<a>` to its parent document and set its `target` as `<iframe>`'s name
4. each referrer policy attribute (eg. HTTP header, <meta>) would contribute to `<a>` directly
5. we call `click()` on `<a>` and when the test was done, we call `postMessage()` inside `<iframe>` to notify its parent document

And target feature for `<a>` and cross origin `postMessage()` is still on its way. My solution is:
1. create a document which is running test harness
2. append an `<iframe>` to its parent document
3. append and `<a>` into `<iframe>`
4. we call `click()` on `<a>` and `<iframe>` navigate to `<a>`'s href

Current solution does not work for some cases:
- HTTP header, it only apply to test harness html document but `<a>` inside `<iframe>`
- cross origin detection, we navigate `<iframe>` rather than its parent document, which make test expectation wrong

One workaround in my mind is to load our test harness html document **inside** `<iframe>` under sandbox, so the test won't run again and we get `<meta>` and HTTP header as we expect. But this would break some consistency in `common.js` and make thing more complex.

---

Sorry for the long description. But I'd like to hear more thought before I actually make things dirty, and find the most proper solution for this.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/13713)

### CODE REVIEW

**Comment by jdm** [comment]
What if instead of loading a-tag.html directly in the iframe, we load a python script which reads a-tag.html and optionally provides the Referrer-Policy header as well?

File: tests/wpt/mozilla/tests/mozilla/referrer-policy/generic/subresource/a-tag.html [hunk 1]
```diff
@@ -3,34 +3,44 @@
...
+        document.head.appendChild(meta);
+      }
+
+      var a = document.createElement("a");
       a.href = "/_mozilla/mozilla/referrer-policy/generic/subresource/document.py?id=" + id + "&redirection=" + redirection +
                "&cache_destroyer=" + cache_destroyer;
 
       for (var attr in referrer_policy) {
-        a[attr] = referrer_policy[attr];
+        // TODO crashed when you assigned value to rel attribute
```
Comment (jdm) [inline tests/wpt/mozilla/tests/mozilla/referrer-policy/generic/subresource/a-tag.html]: I think this isn't necessary any more after #12799 was fixed.

File: tests/wpt/mozilla/tests/mozilla/referrer-policy/generic/subresource/a-tag.html [hunk 2]
```diff
@@ -3,34 +3,44 @@
...
       var referrer_policy;
       try {
-        referrer_policy = JSON.parse(tag_attrs);  
+        referrer_policy = JSON.parse(tag_attrs);
       } catch(e) {
         referrer_policy = {};
       }
 
-      var a = document.getElementById("button");
+      if (meta_referrer !== "undefined") {
```
Comment (jdm) [inline tests/wpt/mozilla/tests/mozilla/referrer-policy/generic/subresource/a-tag.html]: We could move all of the dynamic page changes into the python script if we used it (just do string replacement on stuff like `<!-- a tag -->`. Not sure if that's better or not.

File: tests/wpt/mozilla/tests/mozilla/referrer-policy/generic/common.js [hunk 3]
```diff
@@ -252,7 +263,7 @@ function queryLink(url, callback, referrer_policy) {
       for (var attr in referrer_policy) {
         // TODO crashed when you assigned value to rel attribute
```
Comment (jdm) [inline tests/wpt/mozilla/tests/mozilla/referrer-policy/generic/common.js]: I don't think this is true after #12799 was fixed.

**Comment by jdm** [comment]
I'm not clear on why the cross-origin situation isn't working - I thought that was why we were using the python stash to communicate, since cross-origin postMessage doesn't exist yet.

File: tests/wpt/mozilla/tests/mozilla/referrer-policy/generic/subresource/a-tag.html (hunk 2 above)
Comment (TheKK) [inline tests/wpt/mozilla/tests/mozilla/referrer-policy/generic/subresource/a-tag.html]: This sounds like a reasonable solution. I'll give it a try!

**Comment by TheKK** [comment]
> I'm not clear on why the cross-origin situation isn't working

Say that we have:
  - resource A: insecure-protocol.no-redirect.http.html
  - resource B: a-tag.html
  - resource C: document.py

Normally, we read `resource A` in and some element like `<link>` load `resource C`. For `resource C` its referrer would be `resource A`.

But for tests of `<a>` this isn't true, since we load `resource B` into iframe then navigate to `resource C`. Its referrer becomes `resource B` instead of `resource A`.

**Comment by jdm** [comment]
Oh, I see. Which part of the test harness gets confused by that?

**Comment by TheKK** [comment]
As I know, some tests for `<a>` with **origin-when-cross-origin** would be confuse

**Comment by jdm** [comment]
Right, but which part of the test harness code receives the wrong URL as the reference? Is there some way we could modify the tests to pass a different URL for the `<a>` tests?

**Comment by TheKK** [comment]
Maybe we can modify 

```javascript
      /* referrer-policy-test-case.js */

      /* eg. http://www1.web-platform.test:8000/_mozilla/mozilla/referrer-policy/generic/subresource/a-tag.html?redirection=no-redirect&cache_destroyer=1476561407449 */
      t._subresourceUrl = t._scenario.target_protocol + "://" +
                          domainForOrigin[t._scenario.origin] +
                          normalizePort(targetPort) +
                          t._scenario["subresource_path"] +
                          "?redirection=" + t._scenario["redirection"] +
                          "&cache_destroyer=" + (new Date()).getTime();
```

and

```javascript
      /* a-tag-html */
      var a = document.createElement("a");
      a.href = "/_mozilla/mozilla/referrer-policy/generic/subresource/document.py" +
        "?id=" + id +
        "&redirection=" + redirection +
        "&cache_destroyer=" + cache_destroyer;
```

to delay the origin changing for `<a>` test. I'll try this and see how it going tomorrow.

**Comment by TheKK** [comment]
Is there any way I can inspect the referrer policy when it was delivered by HTTP header to test harness document? I successfully delivered referrer policy via HTTP header to loading `document.py` but I can't infer the policy the test like to apply.

**Comment by jdm** [comment]
I do not believe so. Maybe we could add a meta element to every test that contains the test's referrer policy?

**Comment by nox** [comment]
@jdm @TheKK So what do we do here?

**Comment by jdm** [comment]
I'll take a look at addressing the previous question sometime in the next few days.

**Comment by jdm** [comment]
I have a commit that makes the HTTP header tests pass at https://github.com/jdm/servo/commit/0819c96ff8825a3a62c44d176f46e2bd40d63c74. @TheKK Are you still interested in finishing up this work?

**Comment by TheKK** [comment]
Sure! Give me some time to recap my work :)

**Comment by nox** [comment]
@TheKK Ping? :)

**Comment by TheKK** [comment]
@nox sorry that I only had a little spare time working on this issue in the last busy weeks. Since the storm has passed I can finally resume my work next day :)

**Comment by nox** [comment]
No problem, just a friendly ping.

**Comment by TheKK** [comment]
@jdm Do you mind me apply your change directly?

Since I left this PR for 5 months, I'd like to double check I'm on the correct path:

- all test with referrer policy deliver by `content attribute` should fail for now. (that's why I'm working on #11861)
- all test with referrer policy deliver by `meta-csp` should fail.
- all test with referrer policy deliver by `http header`, `meta-referrer` and `relation` should pass.

Technically this issue has been fixed thanks to jdm's help. I'll clean up mess, add some comments if necessary and make sure that I won't introducing any defects this time. And hopefully, this PR should be able to be merged after that. :)

**Comment by jdm** [comment]
That summary looks correct to me. Please go ahead and merge my changes with your existing ones!

**Comment by TheKK** [comment]
I've clean my previous changes and squashed them into single commit. It should be ready to be reviewed.

One thing that confuses me is the naming style to use in `common.js ` because there exists both snake_case and camelStyle. Please let me know If camel style is proper choice so that I can fix that.

File: tests/wpt/mozilla/tests/mozilla/referrer-policy/generic/subresource/a-tag.html [hunk 4]
```diff
@@ -3,34 +3,58 @@
...
+      var a = document.createElement("a");
+      a.href = protocol + "//" + host +
+        "/_mozilla/mozilla/referrer-policy/generic/subresource/document.py" +
+        "?id=" + id +
+        "&redirection=" + redirection +
+        "&cache_destroyer=" + cache_destroyer;
 
       for (var attr in referrer_policy) {
-        a[attr] = referrer_policy[attr];
+        if (attr === "rel") {
```
Comment (jdm) [inline tests/wpt/mozilla/tests/mozilla/referrer-policy/generic/subresource/a-tag.html]: We should not need this special-case any longer.

File: tests/wpt/mozilla/tests/mozilla/referrer-policy/generic/common.js (hunk 3 above)
Comment (jdm) [inline tests/wpt/mozilla/tests/mozilla/referrer-policy/generic/common.js]: Let's get rid of this special-case.

File: tests/wpt/mozilla/tests/mozilla/referrer-policy/generic/subresource/mozresource.py [hunk 5]
```diff
@@ -83,6 +84,9 @@ def respond(request,
     if access_control_allow_origin != None:
         response.writer.write_header("access-control-allow-origin",
                                      access_control_allow_origin)
+    if http_header_referrer_policy != None:
+        response.writer.write_header("Referrer-Policy",
```
Comment (jdm) [inline tests/wpt/mozilla/tests/mozilla/referrer-policy/generic/subresource/mozresource.py]: Let's make this `referrer-policy` to match the style of the other headers in this file.

**Review by jdm — CHANGES_REQUESTED** [review]
Three small changes; otherwise this looks good to go!

**Comment by TheKK** [comment]
Opps, seem like that `rel` attribute on `<a>` has not yet been implemented. I'll summit another PR to fix that.

**Comment by jdm** [comment]
@bors-servo: delegate=TheKK
This can merge after #16170 merges and the test expectations are updated appropriately.

**Comment by jdm** [comment]
Ah, this needs `./mach update-manifest`.

**Comment by TheKK** [comment]
totally forgot that...

**Comment by TheKK** [comment]
> {"status": "CRASH", "stack": null, "subtest": null, "action": "test_result", "test": "/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_popups_nonescaping-2.html", "line": 199059, "message": null, "expected": "ERROR"}

hmm, I rebased to latest commit, recompiled, tried this test on my machine and got "ERROR" which is expected result. And idea?

**Comment by jdm** [comment]
That's just #14323 and can be ignored.

**Pull request merged**